docs/async/async-in-depth.md

```diff
@@ -0,0 +1,151 @@
+In the second example above, a `Task<T>` object will be returned from `GetStringAsync`. The use of the `await` keyword will result in control being yielded to the caller as well as a newly created task object, required to monitor progress of the `GetFirstCharactersCountAsync` method. The `await` keyword also unwraps the return value from the task object, from [Task.Result](https://msdn.microsoft.com/en-us/library/dd321468(v=vs.110).aspx).
+
+After the System API call, the request is now in kernel space, making its way to the networking subsystem of the OS (such as `/net` in the Linux Kernel).  Here the OS will handle the networking request *asynchronously*.  Details may be different depending on the OS used (the device driver call may be scheduled as a signal sent back to the runtime, or a device driver call may be made and *then* a signal sent back), but eventually the runtime will be informed that the networking request is in progress.  At this time, the work for the device driver will either be scheduled, in-progress, or already finished (the request is already out "over the wire") - but because this is all happening asynchronously, the device driver is able to immediately handle something else!
+
+For example, in Windows an OS thread makes a call to the network device driver and asks it to perform the networking operation via an Interrupt Request Packet (IRP) which represents the operation.  The device driver recieves the IRP, makes the call to the network, marks the IRP as "pending", and returns back to the OS.  Because the OS thread now knows that the IRP is "pending", it doesn't have any more work to do for this job and "returns" back so that it can be used to perform other work.
+
+![](../../images/async/io-1.jpg)
+(caption) A view into what happens when an async call is made.
+
+When the request is fulfilled and data comes back through the device driver, it notifies the CPU of new data received via an interrupt.  How this interrupt gets handled will vary depending on the OS, but eventually the data will be passed through the OS until it reaches a system interop call (for example, in Linux an interrupt handler will schedule the bottom half of the IRQ to pass the data up through the OS asynchronously).  Note that this *also* happens asynchronously!  The result is queued up until the next available thread is able execute the async method and "unwrap" the result of the completed task.
+
+![](../../images/async/io-2.jpg)
+(caption) A view into what happens when an I/O request is complete.
+
+Throughout this entire process, a key takeaway is that **no thread is dedicated to running the task**.  Tasks have no thread affinity.  Although work is executed in some contexts (i.e. the OS does have to pass data to a device driver and respond to an interrupt), there is no thread dedicated to *waiting* for data from the request to come back.  This allows the system to handle a much larger volume of work rather than waiting for some I/O call to finish.
```

Comment:
I'd disagree with the "Tasks have no thread affinity" statement. That can be controlled with Task.ConfigureAwait(), but by default, Task continuations will execute on a thread of the same type as the thread that originated the task.